### PR TITLE
Added the 2D mode feature

### PR DESCRIPTION
**Pull request by julienbernat**
Added orthogonal projection but there is still the problem with the image zooming when going into orthogonal projection. I'm blocked on that but the rest works just fine.

### CODE REVIEW

**Comment by CHrlS98** [comment]
@julienbernat Can you rebase/merge with `upstream`?

**Comment by CHrlS98** [comment]
@julienbernat Also,

1. Can you make it so that the entire content of the options fits by changing the default window size for the slicer window (attachment below)? 
2. Can you make it so that the selected axis matches the individual slicers? For example, when locked in "Z" I'd like to be able to change the current slice I'm looking at by moving the "Z-slice" slider.
3. Can you try changing the projection mode to orthogonal when in 2D mode?
4. It would also be important you test with real data because some slices are not oriented correctly when working on a full brain.

image

**Comment by CHrlS98** [comment]
Some comments on the behaviour of the application:
1. First, well played. Nice first pass!
2. We need to rework the zoom feature, at least for the orthogonal projection; modifying the camera position makes it so that we sometimes end up behind a slice of interest for a given zoom level. I think the camera position should stay still when in 2D mode. 
3. Can we hide the glyphs belonging to other slices than the one of interest when in 2D?

@julienbernat 
How do you wanna do it? We fix the current PR or keep as is and add issues for a later PR?

File: Engine/include/camera.h [hunk 1]
```diff
@@ -72,6 +78,9 @@ class Camera
     /// Window aspect ratio.
     float mAspect;
 
+    /// Camera model is orthogonal.
+    bool isOrthogonal;
```
Comment (CHrlS98) [inline Engine/include/camera.h:82]: Class members should start with `m` -> `mIsOrthogonal`.

File: Engine/include/coordinate_system.h [hunk 2]
```diff
@@ -28,6 +28,10 @@ class CoordinateSystem
     /// \param[in] t Transform to be appplied.
     void ApplyTransform(const glm::mat4& t);
 
+    /// Apply a matrix to the coordinate system.
+    /// \param[in] t Matrix to be appplied.
+    void ApplyMatrix(const glm::mat4& t);
```
Comment (CHrlS98) [inline Engine/include/coordinate_system.h:33]: "Apply" sounds like it is a transformation on top of the current transform (like `ApplyTransform` (l.29)).

May I suggest to rename to `ResetMatrix`, maybe?

File: Engine/include/scene.h [hunk 3]
```diff
@@ -43,7 +55,13 @@ class Scene
...
     std::shared_ptr<ApplicationState> mState;
 
+    /// Reference to the Camera.
+    std::shared_ptr<Camera> mCamera;
+
     /// Vector of models to be rendered.
     std::vector<std::shared_ptr<Model>> mModels;
+    
+    ///Boolean to block the rotation of the scene
+    bool blockRotation;
```
Comment (CHrlS98) [inline Engine/include/scene.h:65]: Start with `m`.

File: Engine/include/scene.h [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #include <vector>
 #include <coordinate_system.h>
 #include <application_state.h>
+#include <camera.h>
```
Comment (CHrlS98) [inline Engine/include/scene.h:6]: Why does the scene need a reference to the camera? Couldn't the camera register a callback on the state viewmode and update itself?

File: Engine/src/gui.cpp [hunk 5]
```diff
@@ -122,7 +122,7 @@ void UIManager::drawSlicersWindow()
         return;
 
     ImGui::SetNextWindowPos(ImVec2(5.f, 25.f), ImGuiCond_FirstUseEver);
-    ImGui::SetNextWindowSize(ImVec2(417.f, 200.f), ImGuiCond_FirstUseEver);
+    ImGui::SetNextWindowSize(ImVec2(417.f, 240.f), ImGuiCond_FirstUseEver);
```
Comment (CHrlS98) [inline Engine/src/gui.cpp:125]: If you increase the value just a little bit, the scroll bar will disappear and that'll look better.

File: Engine/src/gui.cpp [hunk 6]
```diff
@@ -184,6 +184,24 @@ void UIManager::drawSlicersWindow()
         mState->VoxelGrid.SliceIndices.Update(slice);
     }
 
+    ImGui::Text("2D mode");
+    ImGui::SameLine();
+    static int e = 0;
```
Comment (CHrlS98) [inline Engine/src/gui.cpp:189]: Give a more explicit name to the variable `e`.

File: Engine/src/scene.cpp [hunk 7]
```diff
@@ -22,6 +26,46 @@ void Scene::AddSHField()
...
+        const glm::mat4 transform(1.0f);
+        mCoordinateSystem->ApplyMatrix(transform);
+        if(mode == State::ModeEnum::x){
+            glm::mat4 transform = glm::rotate(1.5708f, glm::vec3(0.0, -1.0, 0.0)); 
+            glm::mat4 transform2 = glm::rotate(1.5708f, glm::vec3(0.0, 0.0, -1.0)); 
+            mCoordinateSystem->ApplyTransform(transform);
+            mCoordinateSystem->ApplyTransform(transform2);
+
+        }
+        if(mode == State::ModeEnum::y){
```
Comment (CHrlS98) [inline Engine/src/scene.cpp:60]: `else if` to avoid check if previous condition is true.

File: Engine/src/scene.cpp [hunk 8]
```diff
@@ -22,6 +26,46 @@ void Scene::AddSHField()
...
+            mCamera->ApplyPerspectiveProjection();
+            blockRotation=false;
+            return;
+        }
+        blockRotation=true;
+        mCamera->ApplyOrthogonalProjection();
+        const glm::mat4 transform(1.0f);
+        mCoordinateSystem->ApplyMatrix(transform);
+        if(mode == State::ModeEnum::x){
+            glm::mat4 transform = glm::rotate(1.5708f, glm::vec3(0.0, -1.0, 0.0));
```
Comment (CHrlS98) [inline Engine/src/scene.cpp:54]: For `pi` and `pi/2` definitions, see https://glm.g-truc.net/0.9.4/api/a00145.html.

`transform` and `transform2` can be `const`. Maybe better names for the variables, for example `rotationY` and `rotationZ`?

File: Engine/src/scene.cpp [hunk 9]
```diff
@@ -22,6 +26,46 @@ void Scene::AddSHField()
...
+        }
+    );
+
+}
+
+void Scene::setMode(State::ModeEnum previous, State::ModeEnum mode)
+{
+    if(previous != mode)
+    {
+        if(mode == State::ModeEnum::off){
```
Comment (CHrlS98) [inline Engine/src/scene.cpp:44]: Linebreak before curly braces.

File: Engine/src/scene.cpp [hunk 10]
```diff
@@ -22,6 +26,46 @@ void Scene::AddSHField()
...
+        blockRotation=true;
+        mCamera->ApplyOrthogonalProjection();
+        const glm::mat4 transform(1.0f);
+        mCoordinateSystem->ApplyMatrix(transform);
+        if(mode == State::ModeEnum::x){
+            glm::mat4 transform = glm::rotate(1.5708f, glm::vec3(0.0, -1.0, 0.0)); 
+            glm::mat4 transform2 = glm::rotate(1.5708f, glm::vec3(0.0, 0.0, -1.0)); 
+            mCoordinateSystem->ApplyTransform(transform);
+            mCoordinateSystem->ApplyTransform(transform2);
+
```
Comment (CHrlS98) [inline Engine/src/scene.cpp:58]: Remove newline.

File: Engine/src/scene.cpp [hunk 11]
```diff
@@ -33,12 +77,14 @@ void Scene::Render()
 
 void Scene::RotateCS(const glm::vec2& vec)
 {
-    const float& rotationSpeed = mState->Window.RotationSpeed.Get();
-    const float dx = -vec.x * rotationSpeed;
-    const float dy = -vec.y * rotationSpeed;
-    glm::mat4 transform = glm::rotate(dx, glm::vec3(0.0, 1.0, 0.0));
-    transform = glm::rotate(dy, glm::vec3(1.0, 0.0, 0.0)) * transform;
-    mCoordinateSystem->ApplyTransform(transform);
+    if(!blockRotation){
```
Comment (CHrlS98) [inline Engine/src/scene.cpp:80]: Line break before curly braces.

**Review by CHrlS98 — COMMENTED** [review]
Alright!

Some comments are just regarding coding standards. We should probably add some contributing guidelines for this kind of stuff. I'll draft a PR eventually.

File: Engine/include/scene.h [hunk 12]
```diff
@@ -35,6 +36,16 @@ class Scene
...
+    
+    /// Set the state for the 2d mode
+    /// \param[in] previous Previous value.
+    /// \param[in] mode New value for fading behaviour.
+    void setMode(State::CameraMode previous, State::CameraMode mode);
+
+protected:
+
+    /// \see Model::registerStateCallbacks()
+    void registerStateCallbacks();
```
Comment (CHrlS98) [inline Engine/include/scene.h:48]: can be private

File: Engine/include/scene.h [hunk 13]
```diff
@@ -35,6 +36,16 @@ class Scene
     /// Translate the scene's coordinate system based on mouse move.
     /// \param[in] v Mouse move vector.
     void TranslateCS(const glm::vec2& v);
+    
+    /// Set the state for the 2d mode
+    /// \param[in] previous Previous value.
+    /// \param[in] mode New value for fading behaviour.
+    void setMode(State::CameraMode previous, State::CameraMode mode);
```
Comment (CHrlS98) [inline Engine/include/scene.h:43]: can be private?

File: Engine/include/camera.h [hunk 14]
```diff
@@ -35,13 +35,29 @@ class Camera
...
+    /// Apply Orthogonal Projection.
+    void ApplyOrthogonalProjection();
+
+    /// Apply perspective projection.
+    void ApplyPerspectiveProjection();
+
+    /// Set the projection for the camera
+    /// \param[in] previous Previous value.
+    /// \param[in] mode New value for fading behaviour.
+    void changeProjection(State::CameraMode previous, State::CameraMode mode);
```
Comment (CHrlS98) [inline Engine/include/camera.h:47]: Shouldn't this method be private?

File: Engine/include/camera.h [hunk 15]
```diff
@@ -35,13 +35,29 @@ class Camera
...
     /// \param[in] delta Mouse wheel offset.
     void Zoom(double delta);
 
     /// Update camera attributes on the GPU.
     void UpdateGPU();
 
+protected:
+
+    /// \see Model::registerStateCallbacks()
+    void registerStateCallbacks();
```
Comment (CHrlS98) [inline Engine/include/camera.h:59]: private

File: Engine/src/scene.cpp [hunk 16]
```diff
@@ -33,12 +75,14 @@ void Scene::Render()
 
 void Scene::RotateCS(const glm::vec2& vec)
 {
-    const float& rotationSpeed = mState->Window.RotationSpeed.Get();
-    const float dx = -vec.x * rotationSpeed;
-    const float dy = -vec.y * rotationSpeed;
-    glm::mat4 transform = glm::rotate(dx, glm::vec3(0.0, 1.0, 0.0));
-    transform = glm::rotate(dy, glm::vec3(1.0, 0.0, 0.0)) * transform;
-    mCoordinateSystem->ApplyTransform(transform);
+    if(!mBlockRotation){
```
Comment (CHrlS98) [inline Engine/src/scene.cpp:78]: line break

File: Engine/include/scene.h [hunk 17]
```diff
@@ -17,6 +17,7 @@ class Scene
 
     /// Constructor.
     /// \param[in] state Reference to the ApplicationState.
+    /// \param[in] camera Reference to the Camera.
```
Comment (CHrlS98) [inline Engine/include/scene.h:20]: remove comment

File: Engine/include/scene.h [hunk 18]
```diff
@@ -35,6 +36,16 @@ class Scene
     /// Translate the scene's coordinate system based on mouse move.
     /// \param[in] v Mouse move vector.
     void TranslateCS(const glm::vec2& v);
+    
+    /// Set the state for the 2d mode
+    /// \param[in] previous Previous value.
+    /// \param[in] mode New value for fading behaviour.
```
Comment (CHrlS98) [inline Engine/include/scene.h:42]: update comment

File: Engine/src/camera.cpp [hunk 19]
```diff
@@ -4,6 +4,10 @@
 #include <math.h>
 #include <application_state.h>
 
+namespace {
+    const float FACTOR_ORTHOGONAL_PROJECTION = 0.001f;
```
Comment (CHrlS98) [inline Engine/src/camera.cpp:8]: No indent for namespaces:
```suggestion
const float FACTOR_ORTHOGONAL_PROJECTION = 0.001f;
```

Also break line after `namespace`.

File: Engine/src/camera.cpp [hunk 20]
```diff
@@ -22,7 +26,9 @@ Camera::Camera(const glm::vec3& position,
 ,mCamParamsData(GPU::Binding::camera)
 ,mState(state)
 {
-    mProjectionMatrix = glm::perspective(mFov, mAspect, mNear, mFar);
+    registerStateCallbacks();
+    mIsOrthogonal=false;
```
Comment (CHrlS98) [inline Engine/src/camera.cpp:30]: you can set to false before braces with the other members:
```
Camera::Camera(...)
:mLookAt(lookAt)
, ...
,mCamParamsData(GPU::Binding::camera)
,mState(state)
,mIsOrthogonal(false)
{
    ...
}
```

File: Engine/src/camera.cpp [hunk 21]
```diff
@@ -37,9 +43,61 @@ void Camera::UpdateGPU()
...
+    else
+    {
+        ApplyPerspectiveProjection();
+    }
+}
+
+void Camera::ApplyOrthogonalProjection()
+{
+    mIsOrthogonal = true;    
+    float width = float(mState->Window.Width.Get());
```
Comment (CHrlS98) [inline Engine/src/camera.cpp:88]: `const` width and height

File: Engine/src/camera.cpp [hunk 22]
```diff
@@ -50,5 +108,8 @@ void Camera::Zoom(double delta)
     mPosition = mPosition + direction * (float)delta;
     mLookAt = mLookAt + direction * (float)delta;
     mViewMatrix = glm::lookAt(mPosition, mLookAt, mUpVector);
+    if(mIsOrthogonal){
```
Comment (CHrlS98) [inline Engine/src/camera.cpp:111]: linebreak before curly braces

**Review by CHrlS98 — COMMENTED** [review]
Some more coding standards. Sorry for the extra work, mostly picky stuff.

Public methods start with a capital letter while private and protected methods start with a lower case letter. Can you make sure all new methods are in the right scope? If it isn't used from outside the class, prefer `private`.

We also put scope curly braces on their own line.

**Comment by CHrlS98** [comment]
@julienbernat 
I thought about it some more, and maybe we should just undo the changes relating to the orthographic projection. We can keep the perspective camera in 2D and in 3D and this way the position is always consistent between both modes. 

Supporting orthogonal projection and modifying the method used to zoom in and out could wait for a future PR. This will allow us to move forward faster with the 2D feature which I think will be really useful for analyzing data. Toggling slices visible and invisible can also be done in a future PR as it is already part of an issue anyway. For now, we can just move the annoying perpendicular slices out of the way using the slicers.

If you think you have time to make the slices invisible and modify the zoom that's perfect, but if you're lacking time (we're past midterm after all), maybe we should split in multiple PRs. Let me know, it's up to you!

**Comment by julienbernat** [comment]
@CHrlS98 
I created a branch on my fork that contains the code to add the orthogonal projection if ever we need it in the future. 
On this branch I removed the orthogonal projection and now the 2D mode should be complete.
I will be opening a new pr to remove other slices in 2D mode.

File: Engine/src/camera.cpp [hunk 23]
```diff
@@ -40,9 +40,8 @@ void Camera::UpdateGPU()
 void Camera::Resize(const float& aspect)
 {
     mAspect = aspect;
-    mProjectionMatrix = glm::perspective(mFov, mAspect, mNear, mFar);
```
Comment (CHrlS98) [inline Engine/src/camera.cpp:43]: We need that for the window resize to keep the same aspect ratio, else we get poorly scaled images.

File: Engine/src/sh_field.cpp [hunk 24]
```diff
@@ -235,6 +235,7 @@ void SHField::initializeGPUData()
                                       sizeof(SphereData));
     mGridInfoData = GPU::ShaderData(&gridData, GPU::Binding::gridInfo,
                                     sizeof(GridData));
+
```
Comment (CHrlS98) [inline Engine/src/sh_field.cpp:238]: Don't think we need this change

**Review by CHrlS98 — COMMENTED** [review]
Presque fini!|

File: Engine/src/camera.cpp [hunk 25]
```diff
@@ -42,7 +42,7 @@ void Camera::Resize(const float& aspect)
     mAspect = aspect;
     mProjectionMatrix = glm::perspective(mFov, mAspect, mNear, mFar);
 }
-
```
Comment (CHrlS98) [inline Engine/src/camera.cpp:45]: remove whitespaces

File: Engine/src/sh_field.cpp [hunk 26]
```diff
@@ -234,7 +234,7 @@ void SHField::initializeGPUData()
     mSphereInfoData = GPU::ShaderData(&sphereData, GPU::Binding::sphereInfo,
                                       sizeof(SphereData));
     mGridInfoData = GPU::ShaderData(&gridData, GPU::Binding::gridInfo,
-                                    sizeof(GridData));
+                                    sizeof(GridData));
```
Comment (CHrlS98) [inline Engine/src/sh_field.cpp:237]: remove whitespaces

**Review by CHrlS98 — COMMENTED** [review]
final whitespaces to remove

**Review by CHrlS98 — APPROVED** [review]
Alright good job!

Merci, on merge! :tada:

**Pull request merged**